### ISSUE TITLE
Getting error at end of training: AbstractFeatureResourceE does not exist. [Op:SimpleMLModelTrainer]

### ISSUE THREAD

**Issue by sibyjackgrove** [tensorflow/decision-forests#4]
I am getting the following error when I try a simple model.

```
csv_feature_columns =  ['weekday_weekend'] + weather_columns + building_columns + schedules_columns + encoded_time_columns + ["total_site_electricity_kwh"] 

train_df = pd.read_csv(timeseries_file_path,usecols=csv_feature_columns,nrows=10000)

train_ds = tfdf.keras.pd_dataframe_to_tf_dataset(train_df, label="total_site_electricity_kwh")

model = tfdf.keras.RandomForestModel()
model.fit(train_ds)

```
```
157/157 [==============================] - 6s 18ms/step
---------------------------------------------------------------------------
NotFoundError                             Traceback (most recent call last)
<ipython-input-6-ce1e05e4d2c8> in <module>
      1 # Train a Random Forest model.
      2 model = tfdf.keras.RandomForestModel()
----> 3 model.fit(train_ds)
      4 

~/.conda/envs/tensorflow25/lib/python3.7/site-packages/tensorflow_decision_forests/keras/core.py in fit(self, x, y, callbacks, **kwargs)
    743 
    744     history = super(CoreModel, self).fit(
--> 745         x=x, y=y, epochs=1, callbacks=callbacks, **kwargs)
    746 
    747     self._build(x)

~/.conda/envs/tensorflow25/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py in fit(self, x, y, batch_size, epochs, verbose, callbacks, validation_split, validation_data, shuffle, class_weight, sample_weight, initial_epoch, steps_per_epoch, validation_steps, validation_batch_size, validation_freq, max_queue_size, workers, use_multiprocessing)
   1227           epoch_logs.update(val_logs)
   1228 
-> 1229         callbacks.on_epoch_end(epoch, epoch_logs)
   1230         training_logs = epoch_logs
   1231         if self.stop_training:

~/.conda/envs/tensorflow25/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py in on_epoch_end(self, epoch, logs)
    433     logs = self._process_logs(logs)
    434     for callback in self.callbacks:
--> 435       callback.on_epoch_end(epoch, logs)
    436 
    437   def on_train_batch_begin(self, batch, logs=None):

~/.conda/envs/tensorflow25/lib/python3.7/site-packages/tensorflow_decision_forests/keras/core.py in on_epoch_end(***failed resolving arguments***)
    930     del logs
    931     if epoch == 0:
--> 932       self._model._train_model()  # pylint:disable=protected-access
    933 
    934 

~/.conda/envs/tensorflow25/lib/python3.7/site-packages/tensorflow_decision_forests/keras/core.py in _train_model(self)
    864         guide=guide,
    865         training_config=self._advanced_arguments.yggdrasil_training_config,
--> 866         deployment_config=self._advanced_arguments.yggdrasil_deployment_config,
    867     )
    868 

~/.conda/envs/tensorflow25/lib/python3.7/site-packages/tensorflow_decision_forests/tensorflow/core.py in train(input_ids, label_id, model_id, learner, task, generic_hparms, ranking_group, training_config, deployment_config, guide, model_dir, keep_model_in_resource)
    503       training_config=training_config.SerializeToString(),
    504       deployment_config=deployment_config.SerializeToString(),
--> 505       guide=guide.SerializeToString())
    506 
    507 

~/.conda/envs/tensorflow25/lib/python3.7/site-packages/tensorflow/python/util/tf_export.py in wrapper(*args, **kwargs)
    402           'Please pass these args as kwargs instead.'
    403           .format(f=f.__name__, kwargs=f_argspec.args))
--> 404     return f(**kwargs)
    405 
    406   return tf_decorator.make_decorator(f, wrapper, decorator_argspec=f_argspec)

~/.conda/envs/tensorflow25/lib/python3.7/site-packages/tensorflow_decision_forests/tensorflow/ops/training/op.py in simple_ml_model_trainer(feature_ids, label_id, weight_id, model_id, model_dir, learner, hparams, task, training_config, deployment_config, guide, name)
    510       return _result
    511     except _core._NotOkStatusException as e:
--> 512       _ops.raise_from_not_ok_status(e, name)
    513     except _core._FallbackException:
    514       pass

~/.conda/envs/tensorflow25/lib/python3.7/site-packages/tensorflow/python/framework/ops.py in raise_from_not_ok_status(e, name)
   6895   message = e.message + (" name: " + name if name is not None else "")
   6896   # pylint: disable=protected-access
-> 6897   six.raise_from(core._status_to_exception(e.code, message), None)
   6898   # pylint: enable=protected-access
   6899 

~/.conda/envs/tensorflow25/lib/python3.7/site-packages/six.py in raise_from(value, from_value)

NotFoundError: Resource decision_forests/ 12-in/N27tensorflow_decision_forests3ops23AbstractFeatureResourceE does not exist. [Op:SimpleMLModelTrainer]
```

**Comment by janpfeifer** [comment]
Sorry you bumped into this error. Let me see if we can reproduce it here.

**Comment by achoum** [comment]
Can you print the first few rows of `train_df` with `train_df.head(3)`?

Ps: I was able to get the same error by not having any features in the dataset. i.e. the dataset only contains the label. If this is the issue for you, this is a "terrible error message" type bug :)

**Comment by Vlek** [comment]
I am also getting this issue in both Pop!_OS (with a GTX 1080 Ti) as well as Ubuntu 20.04.2 LTS on a VPS without a videocard.

I was attempting to follow the example code [posted here](https://www.tensorflow.org/decision_forests/tutorials/beginner_colab) with my own dataset with a single column.

**Comment by janpfeifer** [comment]
Is it the case that you have no features (except the label) as @achoum mentioned ? Can you print the first few rows of `train_df` with `train_df.head(3)` ?

**Comment by sibyjackgrove** [comment]
> Can you print the first few rows of `train_df` with `train_df.head(3)`?
> 
> Ps: I was able to get the same error by not having any features in the dataset. i.e. the dataset only contains the label. If this is the issue for you, this is a "terrible error message" type bug :)

@achoum Sure, please find the result of train_df.head(3) below:


weekday_weekend | temp_air | temp_dew | relative_humidity | atmospheric_pressure | wind_direction | wind_speed | build_existing_model.bedrooms | build_existing_model.clothes_dryer | build_existing_model.clothes_washer | ... | clothes_washer_power | dishwasher_power | sleep | total_site_electricity_kwh | cos_hour | sin_hour | cos_day_of_week | sin_day_of_week | cos_month | sin_month
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
0 | -18.85 | -25.85 | 50.463270 | 101399.445 | 305.0 | 5.65 | 3 | 0.37037 | 123.0 | ... | 0.0 | 0.0 | 4.0 | 0.806621 | 1.000000 | 0.000000 | 1.0 | 0.0 | 0.866025 | 0.5
0 | -20.00 | -26.10 | 54.897140 | 101438.484 | 310.0 | 5.70 | 3 | 0.37037 | 123.0 | ... | 0.0 | 0.0 | 4.0 | 0.759627 | 0.962917 | 0.269797 | 1.0 | 0.0 | 0.866025 | 0.5
0 | -20.60 | -26.70 | 54.738663 | 101458.000 | 300.0 | 5.70 | 3 | 0.37037 | 123.0 | ... | 0.0 | 0.0 | 4.0 | 0.757838 | 0.854419 | 0.519584 | 1.0 | 0.0 | 0.866025 | 0.5

<p style="box-sizing: unset; margin: 0px 0px 1em; text-align: left; color: rgba(0, 0, 0, 0.87); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">3 rows × 107 columns</p>

**Comment by achoum** [comment]
Second hypothesis: Is one of the column names contains a comma?

```python
# List the column names containing a comma.
print([column for column in train_df.columns if "," in column])
```

If this is the case, and until we fix it (in next release; eta:next week), you can temporally remove the comma from the feature names with the following code:

```python
...
train_df = pd.read_csv(timeseries_file_path,usecols=csv_feature_columns,nrows=10000)

train_df = train_df .rename(mapper=lambda x : x.replace(",", "<comma>"), axis="columns")

train_ds = tfdf.keras.pd_dataframe_to_tf_dataset(train_df, label="total_site_electricity_kwh")
model = tfdf.keras.RandomForestModel()
model.fit(train_ds)
```

This is already the second bug you helped us identify :) Thanks both !

**Comment by sibyjackgrove** [comment]
> ```python
> print([column for column in train_df.columns if "," in column])
> ```

Yes, it turns out 13 of the column names did have a comma. Thanks for the temporary solution. I will look out for your fix in the code.

**Comment by sibyjackgrove** [comment]
@achoum I am getting this new error on training the model after fixing the commas in the column names.

```
157/157 [==============================] - 6s 17ms/step
---------------------------------------------------------------------------
UnknownError                              Traceback (most recent call last)
<ipython-input-13-046c9cde14d2> in <module>
      2 model = tfdf.keras.RandomForestModel()
      3 #model = tfdf.keras.GradientBoostedTreesModel()
----> 4 model.fit(train_ds)
      5 

~/.conda/envs/tensorflow25/lib/python3.7/site-packages/tensorflow_decision_forests/keras/core.py in fit(self, x, y, callbacks, **kwargs)
    743 
    744     history = super(CoreModel, self).fit(
--> 745         x=x, y=y, epochs=1, callbacks=callbacks, **kwargs)
    746 
    747     self._build(x)

~/.conda/envs/tensorflow25/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py in fit(self, x, y, batch_size, epochs, verbose, callbacks, validation_split, validation_data, shuffle, class_weight, sample_weight, initial_epoch, steps_per_epoch, validation_steps, validation_batch_size, validation_freq, max_queue_size, workers, use_multiprocessing)
   1227           epoch_logs.update(val_logs)
   1228 
-> 1229         callbacks.on_epoch_end(epoch, epoch_logs)
   1230         training_logs = epoch_logs
   1231         if self.stop_training:

~/.conda/envs/tensorflow25/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py in on_epoch_end(self, epoch, logs)
    433     logs = self._process_logs(logs)
    434     for callback in self.callbacks:
--> 435       callback.on_epoch_end(epoch, logs)
    436 
    437   def on_train_batch_begin(self, batch, logs=None):

~/.conda/envs/tensorflow25/lib/python3.7/site-packages/tensorflow_decision_forests/keras/core.py in on_epoch_end(***failed resolving arguments***)
    930     del logs
    931     if epoch == 0:
--> 932       self._model._train_model()  # pylint:disable=protected-access
    933 
    934 

~/.conda/envs/tensorflow25/lib/python3.7/site-packages/tensorflow_decision_forests/keras/core.py in _train_model(self)
    876     # Load and optimize the model in memory.
    877     # Register the model as a SavedModel asset.
--> 878     self._model = tf_op.ModelV2(model_path=model_path, verbose=False)
    879 
    880   def _set_from_yggdrasil_model(self,

~/.conda/envs/tensorflow25/lib/python3.7/site-packages/tensorflow_decision_forests/tensorflow/ops/inference/api.py in __init__(self, model_path, verbose)
    318     self._input_builder.build_from_model_path(model_path)
    319     self._compiled_model = _CompiledSimpleMLModelResource(
--> 320         _DiskModelLoader(model_path))
    321 
    322   def apply(self, features: Dict[Text, Tensor]) -> ModelOutput:

~/.conda/envs/tensorflow25/lib/python3.7/site-packages/tensorflow_decision_forests/tensorflow/ops/inference/api.py in __init__(self, model_loader)
    831         self._init_op = self._initialize()
    832     else:
--> 833       self._init_op = self._initialize()
    834 
    835   def _create_resource(self):

~/.conda/envs/tensorflow25/lib/python3.7/site-packages/tensorflow_decision_forests/tensorflow/ops/inference/api.py in _initialize(self)
    838 
    839   def _initialize(self):
--> 840     return self._model_loader.initialize(self)
    841 
    842 

~/.conda/envs/tensorflow25/lib/python3.7/site-packages/tensorflow_decision_forests/tensorflow/ops/inference/api.py in initialize(self, model)
    876                         (model.resource_handle,)):
    877       init_op = op.SimpleMLLoadModelFromPathWithHandle(
--> 878           model_handle=model.resource_handle, path=model_path)
    879 
    880     ops.add_to_collection(ops.GraphKeys.TABLE_INITIALIZERS, init_op)

~/.conda/envs/tensorflow25/lib/python3.7/site-packages/tensorflow/python/util/tf_export.py in wrapper(*args, **kwargs)
    402           'Please pass these args as kwargs instead.'
    403           .format(f=f.__name__, kwargs=f_argspec.args))
--> 404     return f(**kwargs)
    405 
    406   return tf_decorator.make_decorator(f, wrapper, decorator_argspec=f_argspec)

~/.conda/envs/tensorflow25/lib/python3.7/site-packages/tensorflow_decision_forests/tensorflow/ops/inference/op.py in simple_ml_load_model_from_path_with_handle(model_handle, path, name)
    513       return _result
    514     except _core._NotOkStatusException as e:
--> 515       _ops.raise_from_not_ok_status(e, name)
    516     except _core._FallbackException:
    517       pass

~/.conda/envs/tensorflow25/lib/python3.7/site-packages/tensorflow/python/framework/ops.py in raise_from_not_ok_status(e, name)
   6895   message = e.message + (" name: " + name if name is not None else "")
   6896   # pylint: disable=protected-access
-> 6897   six.raise_from(core._status_to_exception(e.code, message), None)
   6898   # pylint: enable=protected-access
   6899 

~/.conda/envs/tensorflow25/lib/python3.7/site-packages/six.py in raise_from(value, from_value)

UnknownError: TensorFlow: INVALID_ARGUMENT: Unknown item TFE_RECORDIO in class pool N26yggdrasil_decision_forests5model13decision_tree14AbstractFormatE. Registered elements are BLOB_SEQUENCE [Op:SimpleMLLoadModelFromPathWithHandle]
```

**Comment by achoum** [comment]
Thanks for the follow up :)!

This next error is less obvious. I suspect a problem with IO, but I need some more details to figure it out.

Do you mind telling me:

1. If you run the same `model.fit` code twice, does it fail twice?

2. Is it the first model you are training (i.e. the first call to `model.fit`) in the python session, or have you been able to train another model before?

3. Can you train the following toy model without getting the error message:

```python
import pandas as pd
ds = pd.DataFrame({"x":[1,2,3,4], "label":[0,1,0,1]})
model = tfdf.keras.RandomForestModel(num_trees=10)
model.fit(tfdf.keras.pd_dataframe_to_tf_dataset(ds,label="label"))
```

4. Can you share the training logs of the problematic `model.fit`.

If you use ipython, the logs will appear by default. If you use colab, you can run:

```python
!pip install wurlitzer -U
from wurlitzer import sys_pipes

with sys_pipes():
    model.fit(train_ds)
```

5. Near the end of the training logs (previous question), you can find the line `[INFO kernel.cc:856] Export model in log directory: ` followed by a directory path e.g. `/tmp/tmpxsmqm10p`. This is the directory containing the problematic trained model.

Can you share the result of the following command (with your directory path instead of `/tmp/tmpxsmqm10p`):

```python
!ls -l -R /tmp/tmpxsmqm10p
```

Cheers,

**Comment by sibyjackgrove** [comment]
> 1. If you run the same `model.fit` code twice, does it fail twice?

If I run 'model.fit'  by itself, I get the following error:

`InvalidArgumentError: No training examples available. [Op:SimpleMLModelTrainer]
`

> 2. Is it the first model you are training (i.e. the first call to `model.fit`) in the python session, or have you been able to train another model before?

Yes, this is the first model.

> 3. Can you train the following toy model without getting the error message:

Yes, I was able to train it.
 
I found that the problem is when I use a large number of training samples. When I tried a few hundred samples from my dataset, it was a training error. But when I train with 10000 training samples, I get the error. Also, I am using Jupyter lab. But I don't see the logs in the tmp folder. Do I need to use the code you mentioned above to get the logs?

**Comment by achoum** [comment]
Thanks for the extra details.

> 3. I found that the problem is when I use a large number of training samples. When I tried a few hundred samples from my dataset, it was a training error. But when I train with 10000 training samples, I get the error. Also, I am using Jupyter lab.

Interesting. 

> Do I need to use the code you mentioned above to get the logs?

Yes, using the two pieces of code I mentioned above would be very informative for us.

```python
!pip install wurlitzer -U
from wurlitzer import sys_pipes

# Assuming `train_ds` is the training dataset.
with sys_pipes():
    model.fit(train_ds)
```

> But I don't see the logs in the tmp folder.

The logs will be informative.

**Comment by sibyjackgrove** [comment]
@achoum  It seems the problem was due to the wrong task specification (regression instead of default classification). After I made the following modifications, the model is training and return the model without error even with more than 10000 samples. 

```
train_ds = tfdf.keras.pd_dataframe_to_tf_dataset(train_df, label="total_site_electricity_kwh", task=tfdf.keras.Task.REGRESSION)
model = tfdf.keras.RandomForestModel(task = tfdf.keras.Task.REGRESSION)
model.compile(metrics=["mse","mae"])
```

**Comment by achoum** [comment]
That's great. Thanks for the details.

The classification model trained on the regressive value was seeing individual regressive values as separate classes. This is much harder to generalize and this explains the extremely large model size for a small (10k examples) dataset.

I'll add a warning (and possibly an error for the extreme case) when a situation like this is detected. Thanks again.

**Comment by achoum** [comment]
The bugs and the non-intuitive configuration problems have been solved in the 0.1.5 release.

Thanks all :)

**Comment by sibyjackgrove** [comment]
@achoum Thank you for fixing the issues with the new release! Closing this issue.
